### ISSUE TITLE
Secret-sharing the output

### ISSUE THREAD

**Issue by haochenuw** [samee/obliv-c#25]
Again this might not be an issue, but I am wondering if obliv-c supports secret-sharing the output? 

One way I thought is for the garbler to send (x,r) where r is random, and the circuit computes 
f(x,y) -r and reveal to the evaluator. But I don't know how to do this in obliv-c, and it seems a little cumbersome...suggestions are appreciated!

**Comment by jackdoerner** [comment]
Given that obliv-c's revealObliv*() functions can reveal secret data to either party individually, you can construct arbitrary secret-sharings of the output. The only secret sharing scheme for which there is explicit support is XOR-sharing, which can be performed more efficiently using the ocToShared*() and ocFromShared*() functions in copy.oh. It's unlikely any other sharing scheme will be supported by the standard library unless there's a good reason why it can't be supported at the application level instead.

**Comment by haochenuw** [comment]
Sounds good! XOR-sharing is all I need. I will try this out tomorrow when I have access to my linux machine.  Thank you a bunch!

**Comment by haochenuw** [comment]
I looked at the function ocToShared but still don't get how to use it. Could you kindly provide an example usage? For example, in the hamming distance test, how do I secret-share the hamming distance to 2 parties? Should this function be called within the hammingDistance() function, or should it be put inside the main function? Thank you!

**Comment by samee** [comment]
For hammingDistance, it may not really make sense to `ocToShared` family. Those are only useful for very specific workloads of large-scale parallel selections. You really want something much simpler here:

  1. Use `feedOblivInt` to input an extra random integer from party 1.
  2. Compute hamming distance as usual.
  3. XOR the hammingDistance result with the random secret instead of revealing it to everyone.
  4. Reveal the XORed value only to party 2.

Now party 1 has a random integer `r`, while party 2 has `result ^ r`. They form an XOR share of the result. Let us know if this helps.

**Comment by haochenuw** [comment]
Yes it helps a lot! So indeed I am also thinking of secret-sharing the output which is a vector of size around 1000. In that case, does it make sense to use ocToShared? Thank you!

**Comment by samee** [comment]
The best I can say is that try ocToShared and see if it gives you a speed up:)
It's not just the size that matters, but also what you do with it. Right now, we don't implement anything that lets you do any arithmetic or bitwise operation on XOR shares (although you can certainly add them). The only thing it supports is multiplexers, which is useful for sorting, data structures, ORAMs and such. So it depends a lot on what you are trying to trying to do, but right now they serve a very narrow purpose.

**Comment by haochenuw** [comment]
Thanks! I guess my final question is the usage of ocToShared.... I can't quite figure out how to use it from the code. When you have time, could you add an example/update the documentation? Of course It's no rush:)

**Comment by samee** [comment]
You are right, I probably should add those. For now, you can look at testshuffle.{c,h,oc} for an example here: https://github.com/samee/sqrtOram/tree/master/test

Sorry if they are a little hard to follow. I'll add proper documentation when I have time.
